npm/vite-dev-server/src/makeCypressPlugin.ts

```diff
@@ -62,11 +68,45 @@ export const makeCypressPlugin = (
 
       server.middlewares.use(`${base}index.html`, (req, res) => res.end(transformedIndexHtml))
     },
-    handleHotUpdate: () => {
-      // restart tests when code is updated
-      devServerEvents.emit('dev-server:compile:success')
+    handleHotUpdate: ({ server, file }) => {
+      debug('handleHotUpdate - file', file)
+      // get the graph node for the file that just got updated
+      let moduleImporters = server.moduleGraph.fileToModulesMap.get(file)
+      let stopIteration = false
+
+      // until we reached a point where the current module is imported by no other
+      while (moduleImporters && moduleImporters.size) {
+        // as soon as we find one of the specs, we trigger the re-run of tests
+        moduleImporters.forEach((mod) => {
+          if (specs.some((spec) => spec.absolute === mod.file)) {
+            debug('handleHotUpdate - compile success')
+            devServerEvents.emit('dev-server:compile:success')
+            // don't go any further if the refresh is done
+            // NOTE: we cannot do return here since we are inside of a sub-function.
+            // It would move us only one step above.
+            stopIteration = true
```

Comment:
Please use a `for...of` then, because it doesn't make sense to continue iterating through the list of importers